08_irq/poll_button/khmod.c

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: MIT and GPL
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/gpio.h>
+#include <linux/timer.h>
+
+MODULE_DESCRIPTION("Button poll module");
+MODULE_AUTHOR("vadym.kharchuk@globallogic.com");
+MODULE_VERSION("0.1");
+MODULE_LICENSE("Dual MIT/GPL");
+
+#define POLL_TIME_MS	20
+#define BLINK_TIME_MS	1
+
+#define GPIO_NUMBER(port, bit) (32 * (port) + (bit))
+
+#define LED_SD  GPIO_NUMBER(1, 22)
+#define LED_MMC GPIO_NUMBER(1, 24)
+
+/* On-board button.
+ *
+ * LCD/HDMI interface must be disabled.
+ */
+#define BUTTON  GPIO_NUMBER(2, 8)
+
+static int button_gpio = -1;
+
+struct timer_list tim_l;
+struct hrtimer tim_h;
+
+enum hrtimer_restart hr_callback(struct hrtimer *timer)
+{
+	gpio_set_value(LED_MMC, 0);
+	return HRTIMER_NORESTART;
+}
+
+static void button_poll(struct timer_list *data)
+{
+	static bool old_button_state;
+	bool button_state;
+	ktime_t kt;
+
+	button_state = gpio_get_value(BUTTON);
+
+	if (old_button_state != button_state) {
+		gpio_set_value(LED_MMC, 1);
+		kt = ktime_set(0, BLINK_TIME_MS * 1000000);
+		hrtimer_start(&tim_h, kt, HRTIMER_MODE_REL);
+	} else if (!button_state) {
+		gpio_set_value(LED_SD, 1);
+	} else {
+		gpio_set_value(LED_SD, 0);
+	}
+
+	old_button_state = button_state;
```

Comment:
this code could be simplified, aslo no need to rewrite old_button_state if it is already equal.
} else {
    gpio_set_value(LED_SD, button_state ? 0 : 1);
    old_button_state = button_state;
}

Reply:
Yes, right. Thank you!